pretyping/evarsolve.ml

```diff
@@ -1600,11 +1600,31 @@ exception MetaOccurInBodyInternal
 
 let rec invert_definition unify flags choose imitate_defs
                           env evd pbty (evk,argsv as ev) rhs =
-  let aliases = make_alias_map env evd in
-  let evdref = ref evd in
-  let progress = ref false in
   let evi = Evd.find evd evk in
   let sign = evar_filtered_context evi in
+  let rhs = whd_beta env evd rhs (* heuristic *) in
+  let fast =
+    let names = ref Id.Set.empty in
+    let rec is_id_subst ctxt s =
+      match ctxt, SList.view s with
+      | (decl :: ctxt'), Some (c, s') ->
+        let id = get_id decl in
+        names := Id.Set.add id !names;
+        (match c with None -> true | Some c -> isVarId evd id c) && is_id_subst ctxt' s'
+      | [], None -> true
+      | _ -> false
+    in
+      is_id_subst sign argsv &&
```

Comment:
Isn't your new representation of evar subst telling you this? here you are rediscovering it. I was expecting the check to be: is the SList morally made of `length ctx`  `None` nodes.

Reply:
Yes, but you need to compute the names anyways. I think we can try to be even more clever but this requires a more in-depth modification, e.g. storing more data in evar infos.

Reply:
A quick analysis of the calls indicate that there is virtually a 100% hit for the fast path when the substitution is the compact identity. This means that in any case as conjectured above, we would always recompute the set of names even if we get an immediate answer. Thus, first checking for trivial identities wouldn't change the profile.